orca-webhook/src/main/groovy/com/netflix/spinnaker/orca/webhook/service/WebhookService.groovy

```diff
@@ -47,30 +58,83 @@ class WebhookService {
+
+    // Only do cancellation if we can determine the rest template to use.
+    if (restTemplateData == null) {
+      log.warn("Cannot determine rest template to cancel the webhook")
+      return null
+    }
+    WebhookStage.StageData stageData = restTemplateData.stageData
+    try {
+      log.info("Sending best effort webhook cancellation to ${stageData.cancelEndpoint}")
+      ResponseEntity<Object> response = restTemplateData.exchange()
+      log.debug(
+          "Received status code {} from cancel endpoint {} in execution {} in stage {}",
+          response.statusCode,
+          stageData.cancelEndpoint,
+          stageExecution.execution.id,
+          stageExecution.id
+      )
+    } catch (HttpStatusCodeException e) {
+      log.warn("Failed to cancel webhook ${stageData.cancelEndpoint} with statusCode=${e.getStatusCode().value()}", e)
+    } catch (Exception e) {
+      log.warn("Failed to cancel webhook ${stageData.cancelEndpoint}", e)
     }
 
-    URI validatedUri = userConfiguredUrlRestrictions.validateURI(
-      restTemplateProvider.getTargetUrl(url)
-    )
-    HttpHeaders headers = buildHttpHeaders(customHeaders)
-    HttpEntity<Object> httpEntity = new HttpEntity<>(headers)
-    return restTemplateProvider.getRestTemplate(url).exchange(validatedUri, HttpMethod.GET, httpEntity, Object)
+  }
+
+  private RestTemplateData getRestTemplateData(String taskType, StageExecution stageExecution) {
+    for (RestTemplateProvider provider : restTemplateProviders) {
+      WebhookStage.StageData stageData = stageExecution.mapTo(provider.getStageDataType())
+      String destinationUrl
+      HttpHeaders headers = buildHttpHeaders(stageData.customHeaders)
+      HttpMethod httpMethod = HttpMethod.GET
+      HttpEntity<Object> payloadEntity = null
+      switch (taskType) {
+        case 'CreateWebhook':
+          destinationUrl = stageData.url
+          payloadEntity = new HttpEntity<>(stageData.payload, headers)
+          httpMethod = stageData.method
+          break
+        case 'MonitorWebhook':
+          destinationUrl = stageData.statusEndpoint
+          payloadEntity = new HttpEntity<>(null, headers)
+          break
+        case 'CancelWebhook':
+          destinationUrl = stageData.cancelEndpoint
+          payloadEntity = new HttpEntity<>(stageData.cancelPayload, headers)
+          httpMethod = stageData.cancelMethod
+          break
+        default:
+          destinationUrl = ''
+          break
+      }
+
+      // Return on the first match
+      if (destinationUrl != null && !destinationUrl.isEmpty() && provider.supports(destinationUrl, stageData)) {
+        URI validatedUri = userConfiguredUrlRestrictions.validateURI(provider.getTargetUrl(destinationUrl, stageData))
+        RestTemplate restTemplate = provider.getRestTemplate(destinationUrl)
+        return new RestTemplateData(restTemplate, validatedUri, httpMethod, payloadEntity, stageData)
+      }
+    }
```

Comment:
I think it's worth adding a log (in addition to the exception you already have when no rest template found) and then you can also log the url/stage params, etc. for better debuggability.
Also, explicitly return `null` to make it less groovy magic